pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1192 @@
+        values = self._digitize(src=self.waveform_source)
+        self._header_sanity_checks(values)
+        self._footer_sanity_checks(values)
+        self._npoints_sanity_checks(values)
+        preamble = {"source": self.waveform_source}
+        return values[self._header_size:-self._footer_size], self._fill_yaxis_preamble(preamble)
+
+    def _acquire_data(self, requested_points=0, sparsing=1):
+        """ Acquire raw data points from the scope. The header, footer and number of points are
+        sanity-checked but they are not processed otherwise. For a description of the input
+        arguments refer to the download_data method.
+        If the number of expected points is big enough, the transmission is splitted in smaller
+        chunks of 20k points and read one chunk at a time. I do not know the reason why,
+        but if the chunk size is big enough the transmission does not complete successfully.
+        :return: raw data points as numpy array and waveform preamble
+        """
+        # Setup waveform acquisition parameters
+        self.waveform_sparsing = sparsing
+        self.waveform_points = requested_points
+        self.waveform_first_point = 0
+
+        # Calculate how many points are to be expected
+        sample_points = self.acquisition_sample_size(self.waveform_source)
+        if requested_points > 0:
+            expected_points = min(requested_points, int(sample_points / sparsing))
+        else:
+            expected_points = int(sample_points / sparsing)
+
+        # If the number of points is big enough, split the data in small chunks and read it one
+        # chunk at a time. For less than a certain amount of points we do not bother splitting them.
+        chunk_bytes = 20000
+        chunk_points = chunk_bytes - self._header_size - self._footer_size
+        iterations = -(expected_points // -chunk_points)
+        i = 0
+        data = []
+        while i < iterations:
+            # number of points already read
+            read_points = i * chunk_points
+            # number of points still to read
+            remaining_points = expected_points - read_points
+            # number of points requested in a single chunk
+            requested_points = chunk_points if remaining_points > chunk_points else remaining_points
+            self.waveform_points = requested_points
+            # number of bytes requested in a single chunk
+            requested_bytes = requested_points + self._header_size + self._footer_size
+            # read the next chunk starting from this points
+            first_point = read_points * sparsing
+            self.waveform_first_point = first_point
+            # read chunk of points
+            values = self._digitize(src=self.waveform_source, num_bytes=requested_bytes)
+            # perform many sanity checks on the received data
+            self._header_sanity_checks(values)
+            self._footer_sanity_checks(values)
```

Comment:
Do you call somewhere only one sanity check? Otherwise it might be good to combine them in one, such that you splice the values only once.

Reply:
Done